feat(forms): Implement strict types for the Angular Forms package.

dylhunn opened the pull request:
This PR strongly types the forms package by adding generics to `AbstractControl` classes as well as `FormBuilder`. This makes forms type-safe and null-safe, for both controls and values.

The design uses a "control-types" approach. In other words, the type parameter on `FormGroup` is an object containing controls, and the type parameter on `FormArray` is a control.

Issue: #13721

NetanelBasal commented:
There are a lot of challenges in this area due to the flexibility of the API. As far as I can tell, I've written a typed version that reaches the maximum we can without breaking the API. Feel free to look at the source code:

https://github.com/ngneat/reactive-forms

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -795,26 +881,30 @@ export abstract class AbstractControl {
    * Patches the value of the control. Abstract method (implemented in sub-classes).
    */
-  abstract patchValue(value: any, options?: Object): void;
+  abstract patchValue(value: ValT, options?: Object): void;
```
Just an idea: should we call the `ValT` (partial) and `RawValT` (full) pair as `PartialValueT` and `ValueT`? I think that might convey the meaning a bit better.

sonukapoor replied:
Agreed.

dylhunn replied:
Fixed.

AndrewKushnir reviewed (commented):
FYI, just sharing first batch of comments, will publish more comments tomorrow.

criskrzysiu commented:
one question, should new spec follow https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/#exact-optional-property-types  in how optional fields are specified?

For example if we have model like 
```typescript
{
age?: number
}
````
then the control can not be removed using removeControl()

while with
```typescript
{
age: number | undefined
}
```
this should be possible

ramthir commented:
I tested out the typed form a bit, and it's excellent with inferring types and stuff.

I've one question. When we `patchValue`, we use a `Partial` data structure, but when we `setValue` we're considering all the properties to be mandatory. Of course, we allow it to be null, but can't we make the optional fields based on the mandatory validation optional in the type itself?

dylhunn commented on packages/forms/src/form_builder.ts:
```diff
@@ -20,163 +21,200 @@ function isAbstractControlOptions(options: AbstractControlOptions|
+   control<T>(
+       formState: T|FormState<T>,
+       validatorOrOpts?: ValidatorFn|ValidatorFn[]|AbstractControlOptions|null,
+       asyncValidator?: AsyncValidatorFn|AsyncValidatorFn[]|null): FormControl<T|null>;
```
Discussed in RFC -- controls that don't reset to their initial value will be nullable. This will also be documented at length.

SchnWalter replied:
> Discussed in RFC

Where can I find the RFC? I couldn't find it.

(Thanks for all the hard work)

dylhunn commented on packages/forms/src/form_builder.ts:
```diff
@@ -20,163 +21,200 @@ function isAbstractControlOptions(options: AbstractControlOptions|
- *
- * @publicApi
+ * ControlConfig<T> is a tuple containing a value of type T, plus optional validators and async
+ * validators.
```
Fixed.

dylhunn commented:
Rerunning failures reveals these are flakes: https://fusion2.corp.google.com/presubmit/tap/416417012/OCL:416417012:BASE:416422956:1639528317533:31818db4/targets

dylhunn commented:
[The RFC with this PR as a demo](https://github.com/angular/angular/discussions/44513
) has now been posted!

dylhunn commented:
This is starting to converge. The [RFC is complete](https://github.com/angular/angular/discussions/44513), a number of preparatory PRs have landed, and this is now rebased on head. The next step is to update this PR with feedback from the RFC.

dylhunn commented:
All bugs that I'm aware of are fixed, and the PR is reduced to its smallest possible footprint. I think this is now ready for review.

atscott commented on packages/forms/test/typed_integration_spec.ts:
```diff
@@ -0,0 +1,817 @@
+  });
+
+  it('should support groups with union types', () => {
+    interface cat {
+      lives: number;
+    }
+    interface person {
+      nickname: string;
+    }
+    interface catControls {
+      lives: FormControl<number>;
+    }
+    interface personControls {
+      nickname: FormControl<string>;
+    }
```
```suggestion
    interface Cat {
      lives: number;
    }
    interface Person {
      nickname: string;
    }
    interface CatControls {
      lives: FormControl<number>;
    }
    interface PersonControls {
      nickname: FormControl<string>;
    }
```
nit: code style

dylhunn replied:
Fixed.

atscott commented on packages/forms/src/model.ts:
```diff
@@ -211,6 +211,88 @@ function assertAllValuesPresent(control: FormGroup|FormArray, value: any): void
+
+/**
+ * Tokenize splits a string literal S by a delimeter D.
+ */
+type Tokenize<S extends string, D extends string> =                 /*\n*/
+    string extends S ? string[] : /* S must be a literal */         /*\n*/
+                       S extends `${infer T}${D}${infer U}` ? [T, ...Tokenize<U, D>] : /*\n*/
+                                           [S] /* Base case */                                             /*\n*/
+    ;
+
+/**
+ * CoerceStrArrToNumArr accepts an array of strings, and converts any numeric string to a number.
+ */
+type CoerceStrArrToNumArr<S> =                                           /*\n*/
+    S extends [infer Head, ...infer Tail] ?                              /*\n*/
+    Head extends `${number}` ? [number, ...CoerceStrArrToNumArr<Tail>] : /*\n*/
+                               [Head, ...CoerceStrArrToNumArr<Tail>] :                              /*\n*/
+    []                                                                   /*\n*/
+    ;
+
+/**
+ * Navigate takes a type T and an array K, and returns the type of T[K[0]][K[1]][K[2]]...
+ */
+type Navigate<T, K extends(Array<string|number>)> =                                           /*\n*/
+    T extends object ? /* T must be indexable (object or array) */                            /*\n*/
+    (K extends [infer Head, ...infer Tail] ? /* Split K into head and tail */                 /*\n*/
+         (Head extends keyof T ? /* head(K) must index T */                                   /*\n*/
+              (Tail extends(string|number)[] ? /* tail(K) must be an array */                 /*\n*/
+               [] extends Tail ? T[Head] : /* base case: K can be split, but Tail is empty */ /*\n*/
+                   (Navigate<T[Head], Tail>) /* explore T[head(K)] by tail(K) */ :            /*\n*/
+               any) /* tail(K) was not an array, give up */ :                                 /*\n*/
+              any) /* head(K) does not index T, give up */ :                                  /*\n*/
+         any) /* K cannot be split, give up */ :                                              /*\n*/
+    any /* T is not indexable, give up */                                                     /*\n*/
+    ;
```
Would it be more helpful to just turn clang format off for these lines with `// clang-format: off`/`// clang-format: on` comments?

dylhunn replied:
Yes, I've made this change.

atscott reviewed (dismissed):
reviewed-for: public-api

dylhunn commented on goldens/public-api/forms/forms.md:
```diff
@@ -278,13 +278,22 @@ export class FormArrayName extends ControlContainer implements OnInit, OnDestroy
+    }>;
     // @deprecated
-    group(controlsConfig: {
+    group(cc: {
```
I believe I renamed this parameter because `controlsConfig` was no longer accurate -- the proper type is wider. I went ahead and used `childControls` wherever applicable.

dylhunn commented on packages/forms/test/typed_integration_spec.ts:
```diff
@@ -0,0 +1,817 @@
+
+import {FormArray, FormControl, FormGroup} from '../src/forms';
+
+(function() {
```
I don't think so, I must have copied this from elsewhere. I have removed it.

dylhunn commented on packages/forms/test/typed_integration_spec.ts:
```diff
@@ -0,0 +1,817 @@
+describe('FormControl', () => {
+  it('should support inferred controls', () => {
+    const c = new FormControl('', {initialValueIsDefault: true});
+    {
```
The stuff inside `{ ... }` is identical in every test, and is a multiline check I couldn't figure out how to factor out. I can change it if you like.

dylhunn commented on packages/forms/test/typed_integration_spec.ts:
```diff
@@ -0,0 +1,817 @@
+      type ValueType = string|null;
+      let t: ValueType = c.value;
+      let t1 = c.value;
+      t1 = null as unknown as ValueType;
```
Technically yes, but I've used precisely the same format in every test as it's a bit tricky to check assignability both ways.

dylhunn commented on packages/forms/src/form_builder.ts:
```diff
@@ -20,162 +20,195 @@ function isAbstractControlOptions(options: AbstractControlOptions|
+  * `FormControl`, `FormGroup`, or `FormArray`. It reduces the amount of boilerplate needed to
+  * build complex forms.
+  *
+  * @see [Reactive Forms Guide](/guide/reactive-forms)
```
Fixed.

dylhunn commented on packages/forms/src/form_builder.ts:
```diff
@@ -20,162 +20,195 @@ function isAbstractControlOptions(options: AbstractControlOptions|
+    * Constructs a new `FormArray` from the given array of configurations,
+    * validators and options.
+    *
+    * TODO: Do we want to support providing default values to create non-nullable controls?
```
Removed -- this TODO is months stale.

AndrewKushnir commented on goldens/public-api/forms/forms.md:
```diff
@@ -225,37 +225,37 @@ export interface Form {
-export class FormArray extends AbstractControl {
-    constructor(controls: AbstractControl[], validatorOrOpts?: ValidatorFn | ValidatorFn[] | AbstractControlOptions | null, asyncValidator?: AsyncValidatorFn | AsyncValidatorFn[] | null);
-    at(index: number): AbstractControl;
+export class FormArray<ElemT extends AbstractControl<Value<ElemT>, RawValue<ElemT>> = any> extends AbstractControl<Array<Value<ElemT>>, Array<RawValue<ElemT>>> {
```
As an idea: would it make sense to call it `ControlT`, so it's a bit better fit for methods (like `at`, `setControl`, etc)?

sonukapoor replied:
I agree `ControlT` would be better here and in the below where `ElemT` is used.

AndrewKushnir commented on goldens/public-api/forms/forms.md:
```diff
@@ -278,13 +278,22 @@ export class FormArrayName extends ControlContainer implements OnInit, OnDestroy
-    array(controlsConfig: any[], validatorOrOpts?: ValidatorFn | ValidatorFn[] | AbstractControlOptions | null, asyncValidator?: AsyncValidatorFn | AsyncValidatorFn[] | null): FormArray;
-    control(formState: any, validatorOrOpts?: ValidatorFn | ValidatorFn[] | FormControlOptions | null, asyncValidator?: AsyncValidatorFn | AsyncValidatorFn[] | null): FormControl;
-    group(controlsConfig: {
+    array<T>(childControls: Array<T | FormState<T> | ControlConfig<T> | AbstractControl<T> | FormControl<T>>, validatorOrOpts?: ValidatorFn | ValidatorFn[] | AbstractControlOptions | null, asyncValidator?: AsyncValidatorFn | AsyncValidatorFn[] | null): FormArray<AbstractControl<T | null>>;
```
I think "childControls" may be a misleading name here. Those controls are elements of an array, not its children. I'd probably call it just "controls" :)

AndrewKushnir commented on goldens/public-api/forms/forms.md:
```diff
@@ -376,39 +386,35 @@ export interface FormControlOptions extends AbstractControlOptions {
-        [key: string]: AbstractControl;
-    }, validatorOrOpts?: ValidatorFn | ValidatorFn[] | AbstractControlOptions | null, asyncValidator?: AsyncValidatorFn | AsyncValidatorFn[] | null);
-    addControl(name: string, control: AbstractControl, options?: {
+export class FormGroup<CtrlT extends {
```
nit: should we use a full name here as well for consistency (`CtrlT` -> `ControlT`)?

AndrewKushnir commented on goldens/public-api/forms/forms.md:
```diff
@@ -376,39 +386,35 @@ export interface FormControlOptions extends AbstractControlOptions {
+    [K in keyof CtrlT]: AbstractControl<Value<CtrlT[K]>, RawValue<CtrlT[K]>>;
+} = {
+    [key: string]: AbstractControl;
+}> extends AbstractControl<FormGroupValue<CtrlT>, FormGroupRawValue<CtrlT>> {
```
I don't see `FormGroupValue` and `FormGroupRawValue` type definitions in the public API (in this file). They'd probably become visible in the public surface, so we should export them too?

AndrewKushnir commented on packages/forms/src/form_builder.ts:
```diff
@@ -20,162 +20,196 @@ function isAbstractControlOptions(options: AbstractControlOptions|
+   group<T extends {[key: string]: any} = any>(
+       childControls: {
+         [K in keyof T]: FormState<T[K]>|ControlConfig<T[K]>|FormControl<T[K]>|
+         AbstractControl<T[K]>|T[K]
```
The `FormControl<T[K]>` is a more precise version of the `AbstractControl<T[K]>`. I believe we need both here, but it's unclear why. Can we add a quick comment to explain this?

AndrewKushnir commented on packages/forms/src/form_builder.ts:
```diff
@@ -20,162 +20,196 @@ function isAbstractControlOptions(options: AbstractControlOptions|
+   }
+
+   control<T = any>(formState: T|FormState<T>, opts: FormControlOptions&{
+     initialValueIsDefault: true
```
Could you please add a comment in the code on why we use `FormControlOptions&{initialValueIsDefault: true}` here?

dylhunn replied:
Done: "This signature captures calls which set `initialValueIsDefault: true`."

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -211,6 +211,92 @@ function assertAllValuesPresent(control: FormGroup|FormArray, value: any): void
+ *
+ * @publicApi
+ */
+export type TypedOrUntyped<T, Typed, Untyped> = T extends never ? Untyped : Typed;
```
If I understand correctly, this type is temporary (while we support mixed typed and untyped forms) and we don't want developers to rely on it? I'd propose adding it into a description.

JoostK replied:
Does this really have to be public API at all? I see it's used in the declaration of the public `FormGroupValue` and `FormGroupRawValue`, but I don't think that necessitates that `TypedOrUntyped` also becomes public as no-one would ever have to refer to that type by name.

Looking closer, do `FormGroupValue` and `FormGroupRawValue` have to be public API? They're used in public API, but I think they can act as opaque type-aliases. Although I guess that would introduce unresolvable references in the documentation?

atscott replied:
FWIW, things have to be marked public api in order to show up on the angular.io site. If it's used in something that appears in the public API and this _wasn't_, it would mean you wouldn't be able to see what `TypedOrUntyped` actually was from aio. This is currently true of the `Tree` and `TreeNode` of the `RouterState` (https://angular.io/api/router/RouterState).

JoostK replied:
Yeah, I thought that would be the case, unfortunately.

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -219,13 +305,18 @@ function assertAllValuesPresent(control: FormGroup|FormArray, value: any): void
  * instantiated directly.
  *
+ * This class has a main type parameter ValueT which represents the value type of the control.
+ * That type can contain partials, due to disabled controls. The optional type parameter RawValueT
```
I think we need to link it to some page that has a thorough description of the disabled state impact on the data structure (that disabled controls are excluded from the `.value`).

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -469,7 +560,8 @@ export abstract class AbstractControl {
    * without passing along {emitEvent: false} as a function argument.
    */
-  public readonly valueChanges!: Observable<any>;
+  // TODO(issue/24571): remove '!'.
```
Drop this `TODO`?

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -1160,12 +1272,6 @@ export abstract class AbstractControl {
   _onDisabledChange: Array<(isDisabled: boolean) => void> = [];
 
-  /** @internal */
-  _isBoxedValue(formState: any): boolean {
```
nit: we use "boxed" as a term in some tests. We can do a separate PR to cleanup it before/after this one.

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -1599,17 +1733,79 @@ export const FormControl: ɵFormControlCtor =
+/**
+ * FormGroupValue extracts the type of `.value` from a FormGroup's inner object type.
+ *
+ * @publicApi
```
Do we expect developers to interact with those types? I understand that they'd "leak" into the public API, but we could add a quick comment to the description that those types should only be used when absolutely needed (avoided in most cases).

AndrewKushnir reviewed (commented):
@dylhunn awesome work! 👍 I've left some comments, but they are not critical.

It'd be great if @alxhub and/or @JoostK can take a look at those changes from more in-depth TypeScript perspective.

dram2040 commented on goldens/public-api/forms/forms.md:
```diff
@@ -64,22 +65,21 @@ export abstract class AbstractControl {
     }): void;
-    get parent(): FormGroup | FormArray | null;
-    abstract patchValue(value: any, options?: Object): void;
+    get parent(): AbstractControl<any> | null;
```
**A1B2C3**
-true> - opinion
¬</¿
?>2.17.97

JoostK commented on packages/forms/src/form_builder.ts:
```diff
@@ -20,162 +20,196 @@ function isAbstractControlOptions(options: AbstractControlOptions|
+       return childControls;
+
+     } else if (Array.isArray(childControls)) {  // ControlConfig Array
+       const value: T|FormState<T> = childControls[0];
+       const validator: ValidatorFn|ValidatorFn[]|null =
+           childControls.length > 1 ? childControls[1]! : null;
+       const asyncValidator: AsyncValidatorFn|AsyncValidatorFn[]|null =
+           childControls.length > 2 ? childControls[2]! : null;
+       return this.control<T>(value, validator, asyncValidator);
```
All type annotations and the non-null assertion operators are unnecessary here, as TS has narrowed `childControls` to be the tuple `ControlConfig` here, because that's the only array type among the constituents of the `childControls` parameter's union type. But feel free to keep them if you think it helps clarify the code.

```suggestion
       const value = childControls[0];
       const validator = childControls.length > 1 ? childControls[1] : null;
       const asyncValidator = childControls.length > 2 ? childControls[2] : null;
       return this.control<T>(value, validator, asyncValidator);
```

JoostK replied:
Hmm that's surprising, IIRC I did try the code I suggested and I don't remember seeing any errors. But happy to keep as is as this is indeed non-trivial to follow.

JoostK commented on packages/forms/src/model.ts:
```diff
@@ -186,7 +186,7 @@ function getRawValue(control: AbstractControl): any {
 }
 
-function assertControlPresent(parent: FormGroup|FormArray, key: string|number): void {
+function assertControlPresent(parent: FormGroup<any>|FormArray<any>, key: string|number): void {
```
Consider using `unknown` to avoid proliferation of `any`s in the codebase (same for `assertAllValuesPresent` below)

JoostK commented on packages/forms/src/model.ts:
```diff
@@ -211,6 +211,92 @@ function assertAllValuesPresent(control: FormGroup|FormArray, value: any): void
 }
 
+/**
+ * TypedOrUntyped allows one of two different types to be selected, depending on whether the Forms
```
```suggestion
 * `TypedOrUntyped` allows one of two different types to be selected, depending on whether the Forms
```

JoostK commented on packages/forms/src/model.ts:
```diff
@@ -211,6 +211,92 @@ function assertAllValuesPresent(control: FormGroup|FormArray, value: any): void
 
+/**
+ * TypedOrUntyped allows one of two different types to be selected, depending on whether the Forms
+ * Class it's applied to is typed or not.
```
nit: since `class` doesn't have a symbol (like it has in Java, for example) I'd opt to use lowercase here:

```suggestion
 * class it's applied to is typed or not.
```

JoostK commented on packages/forms/src/model.ts:
```diff
@@ -211,6 +211,92 @@ function assertAllValuesPresent(control: FormGroup|FormArray, value: any): void
+export type TypedOrUntyped<T, Typed, Untyped> = T extends never ? Untyped : Typed;
+
+/**
+ * Value gives the type of `.value` in an AbstractControl.
```
```suggestion
 * Value gives the type of `.value` in an `AbstractControl`.
```

JoostK reviewed (commented):
Here's an initial batch of comments. I've only now reached the interesting TS types, those will take me some more time to go over :-)

dylhunn commented:
Thanks everyone! It will take me a day or so to respond to all the comments, then I will @ my reviewers and put the PR back in reviewable state.

mikaelboff commented:
Very nice job, waiting for this!!

sonukapoor commented on packages/forms/src/form_builder.ts:
```diff
@@ -20,162 +20,196 @@ function isAbstractControlOptions(options: AbstractControlOptions|
+    * @description
+    * Construct a new `FormGroup` instance.
+    *
+    * @param cc A collection of child controls. The key for each child is the name
```
@JoostK I think @AndrewKushnir mentioned somewhere in that this should maybe just called `controls` instead. Is that what it was renamed to?

JoostK commented on packages/forms/test/typed_integration_spec.ts:
```diff
@@ -0,0 +1,815 @@
+      type ValueType = string;
+      let t: ValueType = c.value;
+      let t1 = c.value;
+      t1 = null as unknown as ValueType;
```
What is this trying to assert?

Would it be helpful to have something like the following to more easily/clearly assert that the desired type is produced?

```ts
function expectType<T>(type: T): void {}

expectType<string>(valueWhichShouldBeOfTypeString);
```

JoostK commented on packages/forms/test/typed_integration_spec.ts:
```diff
@@ -0,0 +1,815 @@
+      let t1 = c.getRawValue();
+      t1 = null as unknown as RawValueType;
+    }
+    c.setValue('');
```
I'd personally opt to use `@ts-expect-error` expectations here for invalid usages of the API, e.g.

```ts
// @ts-expect-error
c.setValue(null);
```

As currently we seem to be confirming that anything valid is accepted, but not that invalid values produce a compilation error.

JoostK reviewed (commented):
Having that `typed_integration_spec.ts` open in IntelliJ seems to kill my CPU:

<img width="646" alt="image" src="https://user-images.githubusercontent.com/123679/153757489-29fd88c5-20d1-4b88-9765-5a44ad794285.png">

the drop is when I closed the editor and then reopened it again, spiking yet again. Not sure if this is specific to IntelliJ or if it's the TS language server that's spiking. I suspect it's IntelliJ (almost certain since the language service is single threaded), would be good to have someone confirm that this doesn't cause issues in VS Code.

sonukapoor commented:
> Having that `typed_integration_spec.ts` open in IntelliJ seems to kill my CPU:
> 
> <img alt="image" width="646" src="https://user-images.githubusercontent.com/123679/153757489-29fd88c5-20d1-4b88-9765-5a44ad794285.png">
> 
> the drop is when I closed the editor and then reopened it again, spiking yet again. Not sure if this is specific to IntelliJ or if it's the TS language server that's spiking. I suspect it's IntelliJ (almost certain since the language service is single threaded), would be good to have someone confirm that this doesn't cause issues in VS Code.

I took a quick look at this in VS Code and don't see anything unusual. VS Code behaves normal.

sonukapoor commented:
> Having that `typed_integration_spec.ts` open in IntelliJ seems to kill my CPU:
> 
> <img alt="image" width="646" src="https://user-images.githubusercontent.com/123679/153757489-29fd88c5-20d1-4b88-9765-5a44ad794285.png">
> 
> the drop is when I closed the editor and then reopened it again, spiking yet again. Not sure if this is specific to IntelliJ or if it's the TS language server that's spiking. I suspect it's IntelliJ (almost certain since the language service is single threaded), would be good to have someone confirm that this doesn't cause issues in VS Code.

I took a quick look at this in VS Code and don't see anything unusual. VS Code behaves normal.

dylhunn reviewed (commented):
@JoostK @AndrewKushnir @atscott Thank you all for the thorough review comments! I have responded to all comments, incorporated the changes, and updated the PR. I'm ready for another round of review when you're available, thanks!

dylhunn committed:
feat(forms): Implement strict types for the Angular Forms package.

This PR strongly types the forms package by adding generics to AbstractControl classes as well as FormBuilder. This makes forms type-safe and null-safe, for both controls and values.

The design uses a "control-types" approach. In other words, the type parameter on FormGroup is an object containing controls, and the type parameter on FormArray is an array of controls.

Special thanks to Alex Rickabaugh and Andrew Kushnir for co-design & implementation, to Sonu Kapoor and Netanel Basal for illustrative prior art, and to Cédric Exbrayat for extensive testing and validation.

BREAKING CHANGE: Forms classes accept a generic.

Forms model classes now accept a generic type parameter. Untyped versions of these classes are available to opt-out of the new, stricter behavior.

JoostK commented on packages/forms/src/form_builder.ts:
```diff
@@ -20,162 +20,196 @@ function isAbstractControlOptions(options: AbstractControlOptions|
+     if (childControls instanceof FormControl) {
+       return childControls as FormControl<T>;
+     } else if (childControls instanceof AbstractControl) {  // A control; just return it
+       return childControls;
```
Narrowing at call sites won't depend on the implementation of a function, only its return type.

Here, `childControls` would have been narrowed to `AbstractControl` which satisfies the return type, so TS accepts this return value. If you were to remove the `childControls instanceof FormControl` check then we'd pass this `AbdtractControl` condition instead, and `childControls` would still be accepted as return value as an `AbstractControl` is still a value return type.

The only difference is that there won't be a return statement that has been narrowed to `FormControl<T>` even though that is specified as a return type, but that's not visible from callers.

dylhunn commented:
Typed Forms TGP Update: By using qualified exports (e.g. `export {UntypedFormGroup as FormGroup}`), it was possible to run a TAP global presubmit on the typed forms prototype: [link to deflake run](https://fusion2.corp.google.com/OCL:429799493:BASE:429799995:1645316597494:8b7b1c22). There is only one failing cluster, with a single underlying root cause.

However, the one root cause is depended on by 20 targets, and it exposes an interesting issue. I have created a minimal reproduction: [playground link](https://www.typescriptlang.org/play?#code/PTAEGUEsFsAcBtIDNIFMAmpoHt2vqAC4CesqAzqEtgE6gBit0A4jdgK6ygCGAdpkhpp+5AHQAoEBGzRUobEiIALCnLyFukeJW405kGnvioAbn0IAaUACN2hUJF4BjeOzwDaoJzISpCqXgoxcXESMghCIV4AcwBpVGJyAHkkAB4AFVBUAA9-EXlrACtUJ0IAPlAAXlAAUVyablLUgGsEhVB0q3JIx2iygG4QvBddORHySgBBa26G0oBhbF5I7HgMqp5eYgqAb3FQA9AzV1QALg7Bw68l2fZS2gAKY-YzjoBKQYBfELC5ADVuCd1jk8ugpjNIo1CItlmx4AAfdj8VAoQLoCqVfaHTIggJg0DTWZQmErNZ8bYAfg6AG0AOTPVC0gC6oHOgRMqBog1CpDkjBoLDYnHSvPQSRoAFVlqLUiS4Z0OqKrFLfuiNnsDtTWsRzrNekzzuTxJ8srk8ZQ5atMlSVaLWYqyOhub8GExWBxYACgTizfkdtTYg5eKBte10kyKedCZCFktSZ8MViDvzBR6RY7xbbHRkrAAFXSESCA1L+wOOENtRTh85el4ZANMsoJqz+7W6noxA2bYgJ7niYbwUZeQcTV0C92cWVx+Wm0GUUtBivEdqW+DVgkQubQ6erE3VVsJdtRaJd6Nb1cJ2fmpNXDdE2Ow1apFMT2DpjCZ6XZ1cKl9Cz2AnWP5lFY5JlLsN7eLwtz3DQDxQaS5DnD+Hw3no0SQN0nKrqkga4vk4AdnECTJGkwFlA8vDcLI5yxG85wmNgkBOsaIRSOkSiYdc0HsLIdAuvh+LUAKoDRP+myYE47CGAEhDwMQoAAO60M0wRSAA6nI0DsN0NhjKMSDsAQvDYPYhDYDYejcM0DiEBI-YlIOejDtwo6Ece8TEK+GQVIJlB-h6oAaoc6GYf4NCrpR1GvHqMT0UcTEsd8khgBxXFQeQvGckQvJXvkAWcKB-CgOg2AUKAJn2LYWjoPZA5DuMlDgNABacgVsA+Xl+LtUFaGoBhWERTuax4b6+KhlWFFUTRoB0QxiVfCEQA).

mary-poppins commented:
You can preview d4a3e01 at https://pr43834-d4a3e01.ngbuilds.io/.
You can preview ba6dae0 at https://pr43834-ba6dae0.ngbuilds.io/.
You can preview 788d6da at https://pr43834-788d6da.ngbuilds.io/.
You can preview a3362d2 at https://pr43834-a3362d2.ngbuilds.io/.
You can preview 0e99ee7 at https://pr43834-0e99ee7.ngbuilds.io/.
You can preview d8ede69 at https://pr43834-d8ede69.ngbuilds.io/.
You can preview 4ada340 at https://pr43834-4ada340.ngbuilds.io/.
You can preview d991f13 at https://pr43834-d991f13.ngbuilds.io/.
You can preview e805c6c at https://pr43834-e805c6c.ngbuilds.io/.
You can preview b40238f at https://pr43834-b40238f.ngbuilds.io/.
You can preview 0223e92 at https://pr43834-0223e92.ngbuilds.io/.
You can preview 0501e05 at https://pr43834-0501e05.ngbuilds.io/.
You can preview 10879c9 at https://pr43834-10879c9.ngbuilds.io/.
You can preview 44a1f0b at https://pr43834-44a1f0b.ngbuilds.io/.
You can preview f3b256a at https://pr43834-f3b256a.ngbuilds.io/.
You can preview 3626aab at https://pr43834-3626aab.ngbuilds.io/.
You can preview 8ec6bf7 at https://pr43834-8ec6bf7.ngbuilds.io/.
You can preview 6f1c40e at https://pr43834-6f1c40e.ngbuilds.io/.
You can preview ccc46df at https://pr43834-ccc46df.ngbuilds.io/.
You can preview b496ad6 at https://pr43834-b496ad6.ngbuilds.io/.
You can preview 5850beb at https://pr43834-5850beb.ngbuilds.io/.
You can preview 06f35f3 at https://pr43834-06f35f3.ngbuilds.io/.
You can preview d71ca33 at https://pr43834-d71ca33.ngbuilds.io/.
You can preview 8139256 at https://pr43834-8139256.ngbuilds.io/.
You can preview 281aa00 at https://pr43834-281aa00.ngbuilds.io/.
You can preview 61a333c at https://pr43834-61a333c.ngbuilds.io/.
You can preview 9b85074 at https://pr43834-9b85074.ngbuilds.io/.
You can preview a3da34d at https://pr43834-a3da34d.ngbuilds.io/.
You can preview b34146f at https://pr43834-b34146f.ngbuilds.io/.
You can preview c7c1a20 at https://pr43834-c7c1a20.ngbuilds.io/.
You can preview 058e1ac at https://pr43834-058e1ac.ngbuilds.io/.

mary-poppins commented:
You can preview 45e2a88 at https://pr43834-45e2a88.ngbuilds.io/.

dylhunn commented:
I have updated the PR to avoid [the issue with tsc](https://github.com/microsoft/TypeScript/issues/48033), and am running [another TGP](https://test.corp.google.com/OCL:432053390:BASE:432082831:1646276600728:40d99410) (looks good so far).

alfaproject commented on goldens/public-api/forms/forms.md:
```diff
@@ -21,7 +21,7 @@ import { SimpleChanges } from '@angular/core';
 
 // @public
-export abstract class AbstractControl {
+export abstract class AbstractControl<ValueT = any, RawValueT extends ValueT = ValueT> {
```
I work with a lot of TS projects from multiple companies and I never seen a `T` suffix like this. Generics from C# and Java always use a prefixed `T` like in `TData`, `TReq`, etc

Microsoft, Facebook, Apollo just to give a few examples always have a prefixed T. Is this style a Google thing? I've tried to look it up but couldn't find anything in Google style guides. It looks super weird o:

Actually, I remembered to look into Angular repo and it looks the 'right thing' is already being done https://github.com/angular/angular/search?q=TExpression

dylhunn replied:
Thanks, we have actually received this feedback in a couple places. I've updated all instances to use the `T` as a prefix instead of a suffix.

dylhunn commented:
I will put this back in draft state for now, since I'm OOO next week. Once I'm back I will push to finalize and review the PR. Thanks everyone!

dylhunn commented:
Marking ready for review again, now that I'm back.

alxhub commented on packages/forms/src/model/abstract_model.ts:
```diff
@@ -145,15 +145,105 @@ export function assertControlPresent(parent: any, isGroup: boolean, key: string|
+ *
+ * @publicApi
+ */
+export type Get<T, K> =                                                   /*\n*/
```
All of these helper types need to be exported from the forms package root, as they're effectively part of the public API surface.

We should also give them more descriptive names, and namespace them with `ɵ` to indicate that they're not meant for user consumption.

dylhunn replied:
For all types which are named in the public API, I have marked them exported, added `ɵ` to their names, and reexported them from `forms.ts`.

Do you think I need to do this for inner types as well, which are never named in the public API but are implementation details of public types? (e.g. `abstract_model.ts#Tokenize`)

atscott replied:
I added the aio preview label so you can check what things look like there. I think the answer is going to be "yes". Anything that appears in the public API types will need to be exported through the index or it won't be navigable on aio.

dylhunn replied:
I believe prefixing with `ɵ` actually causes the symbols to be *dropped* from the public docs, which makes sense. I recommend that we drop the `ɵ` from the top-level exported symbols. WDYT?

dylhunn replied:
Resolved according to discussion with Andrew below (https://github.com/angular/angular/pull/43834#discussion_r832739101).

mary-poppins commented:
You can preview d4a3e01 at https://pr43834-d4a3e01.ngbuilds.io/.
You can preview ba6dae0 at https://pr43834-ba6dae0.ngbuilds.io/.
You can preview 788d6da at https://pr43834-788d6da.ngbuilds.io/.
You can preview a3362d2 at https://pr43834-a3362d2.ngbuilds.io/.
You can preview 0e99ee7 at https://pr43834-0e99ee7.ngbuilds.io/.
You can preview d8ede69 at https://pr43834-d8ede69.ngbuilds.io/.
You can preview 4ada340 at https://pr43834-4ada340.ngbuilds.io/.
You can preview d991f13 at https://pr43834-d991f13.ngbuilds.io/.
You can preview e805c6c at https://pr43834-e805c6c.ngbuilds.io/.
You can preview b40238f at https://pr43834-b40238f.ngbuilds.io/.
You can preview 0223e92 at https://pr43834-0223e92.ngbuilds.io/.
You can preview 0501e05 at https://pr43834-0501e05.ngbuilds.io/.
You can preview 10879c9 at https://pr43834-10879c9.ngbuilds.io/.
You can preview 44a1f0b at https://pr43834-44a1f0b.ngbuilds.io/.
You can preview f3b256a at https://pr43834-f3b256a.ngbuilds.io/.
You can preview 3626aab at https://pr43834-3626aab.ngbuilds.io/.
You can preview 8ec6bf7 at https://pr43834-8ec6bf7.ngbuilds.io/.
You can preview 6f1c40e at https://pr43834-6f1c40e.ngbuilds.io/.
You can preview ccc46df at https://pr43834-ccc46df.ngbuilds.io/.
You can preview b496ad6 at https://pr43834-b496ad6.ngbuilds.io/.
You can preview 5850beb at https://pr43834-5850beb.ngbuilds.io/.
You can preview 06f35f3 at https://pr43834-06f35f3.ngbuilds.io/.
You can preview d71ca33 at https://pr43834-d71ca33.ngbuilds.io/.
You can preview 8139256 at https://pr43834-8139256.ngbuilds.io/.
You can preview 281aa00 at https://pr43834-281aa00.ngbuilds.io/.
You can preview 61a333c at https://pr43834-61a333c.ngbuilds.io/.
You can preview 9b85074 at https://pr43834-9b85074.ngbuilds.io/.
You can preview a3da34d at https://pr43834-a3da34d.ngbuilds.io/.
You can preview b34146f at https://pr43834-b34146f.ngbuilds.io/.
You can preview c7c1a20 at https://pr43834-c7c1a20.ngbuilds.io/.
You can preview 058e1ac at https://pr43834-058e1ac.ngbuilds.io/.
You can preview 30630ce at https://pr43834-30630ce.ngbuilds.io/.
You can preview 45e2a88 at https://pr43834-45e2a88.ngbuilds.io/.
You can preview 69ca188 at https://pr43834-69ca188.ngbuilds.io/.
You can preview 5a5206a at https://pr43834-5a5206a.ngbuilds.io/.
You can preview c76fc16 at https://pr43834-c76fc16.ngbuilds.io/.
You can preview 6dbb737 at https://pr43834-6dbb737.ngbuilds.io/.
You can preview cb2ef62 at https://pr43834-cb2ef62.ngbuilds.io/.
You can preview 0935411 at https://pr43834-0935411.ngbuilds.io/.
You can preview 88a236c at https://pr43834-88a236c.ngbuilds.io/.
You can preview d938471 at https://pr43834-d938471.ngbuilds.io/.
You can preview bc16de0 at https://pr43834-bc16de0.ngbuilds.io/.
You can preview da7b896 at https://pr43834-da7b896.ngbuilds.io/.
You can preview 4b2644f at https://pr43834-4b2644f.ngbuilds.io/.
You can preview f4d1918 at https://pr43834-f4d1918.ngbuilds.io/.
You can preview b6d3604 at https://pr43834-b6d3604.ngbuilds.io/.
You can preview 555c6d4 at https://pr43834-555c6d4.ngbuilds.io/.
You can preview f9ce55d at https://pr43834-f9ce55d.ngbuilds.io/.

mary-poppins commented:
You can preview f2e959d at https://pr43834-f2e959d.ngbuilds.io/.

AndrewKushnir commented on packages/forms/src/model/form_group.ts:
```diff
@@ -281,25 +410,29 @@ export class FormGroup extends AbstractControl {
-      if (this.controls[name]) {
-        this.controls[name].patchValue(value[name], {onlySelf: true, emitEvent: options.emitEvent});
+    (Object.keys(value) as Array<keyof TControl>).forEach(name => {
+      if ((this.controls as any)[name]) {
+        (this.controls as any)[name]!.patchValue(
```
Possible refactoring:
```suggestion
      // ... comment on why we use `as any` here ...
      const control = (this.controls as any)[name];
      control?.patchValue(
```

dylhunn replied:
Fixed.

AndrewKushnir reviewed (commented):
@dylhunn looks great, just left some minor comments.

dylhunn commented on packages/forms/src/model/abstract_model.ts:
```diff
@@ -145,15 +145,105 @@ export function assertControlPresent(parent: any, isGroup: boolean, key: string|
+  });
 }
 
+export type ɵIsAny<T, Y, N> = 0 extends(1&T) ? Y : N;
```
Done.

dylhunn commented on packages/forms/src/model/abstract_model.ts:
```diff
@@ -145,15 +145,105 @@ export function assertControlPresent(parent: any, isGroup: boolean, key: string|
+    (T['setValue'] extends((v: infer R) => void) ? R : never) :
+    never;
+
+// clang-format off
```
Done.

dylhunn commented on packages/forms/src/model/abstract_model.ts:
```diff
@@ -162,13 +252,16 @@ export function assertAllValuesPresent(control: any, isGroup: boolean, value: an
  * instantiated directly.
  *
+ * The first type parameter TValue represents the value type of the control (`control.value`).
+ * The optional type parameter TRawValue  represents the raw value type (`control.getRawValue()`).
```
I don't think that `AbstractControl` is the right place to highlight this distinction -- for some subclasses the two types will be related differently (as in `FormControl`). We do need some more documentation though; I will do a separate PR soon to supplement the docs.

dylhunn commented on packages/forms/src/model/form_array.ts:
```diff
@@ -250,11 +265,14 @@ export class FormArray extends AbstractControl {
     value.forEach((newValue: any, index: number) => {
       assertControlPresent(this, false, index);
-      this.at(index).setValue(newValue, {onlySelf: true, emitEvent: options.emitEvent});
+      (this.at(index) as any).setValue(newValue, {onlySelf: true, emitEvent: options.emitEvent});
```
It's actually not needed! I've removed it.

dylhunn commented on packages/forms/src/form_builder.ts:
```diff
@@ -23,213 +23,247 @@ function isAbstractControlOptions(options: AbstractControlOptions|
+     * @description
+     * Construct a new `FormGroup` instance.
+     *
+     * @deprecated This API is not typesafe and can result in issues with Closure Compiler renaming.
```
I think we should clean this up, but in a separate PR. I'll do a follow up after this lands.

dylhunn commented on goldens/public-api/forms/forms.md:
```diff
@@ -41,7 +41,8 @@ export abstract class AbstractControl {
     get enabled(): boolean;
     readonly errors: ValidationErrors | null;
-    get(path: Array<string | number> | string): AbstractControl | null;
+    get<P extends string | (readonly (string | number)[])>(path: P): AbstractControl<ɵGetProperty<TRawValue, P>> | null;
```
All sites fixed, and also resolving the above relevant comment.

AndrewKushnir reviewed (approved):
👍

alxhub commented on packages/forms/src/model/form_control.ts:
```diff
@@ -309,34 +323,44 @@ export interface ɵFormControlCtor {
    */
-  new(formState: any, validatorOrOpts?: ValidatorFn|ValidatorFn[]|FormControlOptions|null,
-      asyncValidator?: AsyncValidatorFn|AsyncValidatorFn[]|null): FormControl;
+  new<T = any>(value: FormState<T>|T, opts: FormControlOptions&{initialValueIsDefault: true}):
```
Should we have a default `T = any` here? I'm not sure it serves any purpose. Untyped creations will go through the `UntypedFormControl` constructor, so I think we should just have a `<T>` parameter here. Or am I misunderstanding?

dylhunn replied:
I think this is OK to change as well; I've applied the fix.

alxhub commented on packages/forms/src/model/form_control.ts:
```diff
@@ -309,34 +323,44 @@ export interface ɵFormControlCtor {
+  prototype: FormControl<any>;
+}
+
+function isFormState<T>(formState: unknown): formState is FormState<T> {
```
```suggestion
function isFormState(formState: unknown): formState is FormState<unknown> {
```

There's no need for this function to make any assertions about the generic type of the `FormState` - the `<T>` parameters is never passed in this code anyway.

dylhunn replied:
Makes sense, fixed.

alxhub commented on packages/forms/src/model/form_control.ts:
```diff
@@ -385,9 +409,9 @@ export const FormControl: ɵFormControlCtor =
+          formState: TValue|FormState<TValue> = this.defaultValue,
           options: {onlySelf?: boolean, emitEvent?: boolean} = {}): void {
-        this._applyFormState(formState);
+        this._applyFormState(formState as FormState<TValue>| TValue);
```
I think this is an unnecessary cast?

dylhunn replied:
Indeed. I have removed it.

alxhub commented on packages/forms/src/model/form_group.ts:
```diff
@@ -201,7 +324,10 @@ export class FormGroup extends AbstractControl {
    * @returns false for disabled controls, true otherwise.
    */
-  contains(controlName: string): boolean {
+  contains<K extends string&keyof TControl>(controlName: K): boolean;
```
Should we limit the type of `contains`? It may not make sense to disallow checking the existence of arbitrary keys.

dylhunn replied:
I agree that it's probably not helpful for us to constrain `contains` like this. I've removed the constraint and added a unit test for it.

alxhub commented on packages/forms/src/form_builder.ts:
```diff
@@ -23,207 +23,241 @@ function isAbstractControlOptions(options: AbstractControlOptions|
+     * * `updateOn`: The event upon which the control should be updated (options: 'change' | 'blur'
+     * | submit').
+     */
+    group<T extends {[key: string]: any} = {[key: string]: any}>(
```
This is another place where I wonder if we should have default type parameters here. Untyped users will go through the `UntypedFormControl` interface, so I don't think these are useful/desirable?

dylhunn replied:
I think this is *probably* OK, and I have removed all the default type arguments on `FormBuilder`. To be honest, I need to add a lot more typed unit tests of `FormBuilder` -- it's not as tested as the new typed model classes. I'll do this tomorrow, to make sure we're not breaking any use case we haven't considered.

dylhunn commented on packages/forms/src/model/form_control.ts:
```diff
@@ -283,7 +297,7 @@ export interface FormControl extends AbstractControl {
 }
 
-type FormControlInterface = FormControl;
+type FormControlInterface<TValue = any> = FormControl<TValue>;
```
Done.

dylhunn commented on packages/forms/src/model/form_array.ts:
```diff
@@ -110,10 +125,10 @@ export class FormArray extends AbstractControl {
    * @param index Index in the array to retrieve the control. If `index` is negative, it will wrap
-   *     around from the back, and if index is greatly negative (less than `-length`), the result is
-   * undefined. This behavior is the same as `Array.at(index)`.
+   *     around from the back, and if index is greatly negative (less than `-length`), the result
```
Fixed.

dylhunn commented on packages/forms/src/model/form_array.ts:
```diff
@@ -74,7 +74,11 @@ import {AbstractControl, AbstractControlOptions, assertAllValuesPresent, assertC
  * @publicApi
  */
-export class FormArray extends AbstractControl {
+export class FormArray<TControl extends AbstractControl<ɵValue<TControl>, ɵRawValue<TControl>> =
```
Your comment got dropped from the GitHub UI due to me deleting that line, but here it was:

> Does this include unknown in the untyped case? It shouldn't - AbstractControl<unknown> should use the typed API. We should really delegate this to the canonical TypedOrUntyped implementation in abstract_model.

This is a really interesting question. I spent quite a bit of time thinking about this, and I think that having a separate `FormArrayTypedOrUntyped` is both confusing and technically incorrect in the `unknown` case.

I have reworked the type of `TControl` a bit -- I'm now using an identical approach to the one above in `FormGroup.controls`. Specifically, `TControl` defaults to `any`, which can always be disambiguated from the typed case of `AbstractControl<T>`, using the same `TypedOrUntyped` helper as above. It's at the cost of a bit of added complexity in the type of `TControl`, but I think that it is more correct.

dylhunn reviewed (commented):
All comments addressed below.

dylhunn commented on packages/forms/src/model/form_array.ts:
```diff
@@ -8,7 +8,18 @@
+ * @publicApi
+ */
+type FormArrayTypedOrUntyped<T extends AbstractControl<any>, Typed, Untyped> =
+    T extends AbstractControl<infer U>? (unknown extends U ? Untyped : Typed) : never;
```
See response below.

alxhub reviewed (approved):
🎉 nice work, Dylan, I think this is in great shape! Just another couple small comments but otherwise LGTM!

dylhunn commented on packages/forms/src/model/abstract_model.ts:
```diff
@@ -145,15 +145,105 @@ export function assertControlPresent(parent: any, isGroup: boolean, key: string|
+ 
+/**
+* Tokenize splits a string literal S by a delimeter D.
+*/
```
Thanks, good catch. Fixed.

dylhunn commented on packages/forms/src/model/abstract_model.ts:
```diff
@@ -145,15 +145,105 @@ export function assertControlPresent(parent: any, isGroup: boolean, key: string|
+/**
+ * Writeable removes readonly from all keys.
+ */
+type Writeable<T> = {
```
Fixed.

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

cexbrayat commented on packages/forms/test/typed_integration_spec.ts:
```diff
@@ -0,0 +1,1032 @@
+          new FormArray([new FormControl(true, {initialValueIsDefault: true})]);
+
+      // const c2: FormArray<FormControl<boolean>> =
+      //     new FormArray([new FormControl(true, {initialValueIsDefault: true})]);
```
FYI this is commented out, is it on purpose?

dylhunn replied:
No -- cleaned up, thanks!

alxhub reviewed (approved):
Reviewed-for: public-api

dylhunn commented on packages/forms/src/model/form_array.ts:
```diff
@@ -8,7 +8,26 @@
+    ɵTypedOrUntyped<T, Array<ɵValue<T>>, any[]>;
+
+/**
+ * FormGroupRawValue extracts the type of `.getRawValue()` from a FormArray's element type, and
```
Yes, thanks!

atscott commented on goldens/public-api/forms/index.md:
```diff
@@ -370,43 +380,74 @@ export interface FormControlOptions extends AbstractControlOptions {
+    }>, controlName: string): boolean;
+    // (undocumented)
+    controls: ɵTypedOrUntyped<TControl, TControl, {
         [key: string]: AbstractControl;
```
For places where we we are opting in to the untyped behavior, should `AbstractControl` explicitly be written as `AbstractControl<any>` rather than just `AbstractControl` (which, from what I understand, is implicitly `AbstractControl<any>` because of the default type)?

```suggestion
        [key: string]: AbstractControl<any>;
```

This question applies above to the `addControl` method and below to the `setControl`

dylhunn replied:
Yes, I will make this update everywhere the `any` is implicit.

atscott reviewed (approved):
reviewed-for: public API

atscott reviewed (approved):
reviewed-for: public-api

itotallyrock commented on packages/forms/src/model/form_control.ts:
```diff
@@ -145,7 +155,7 @@ export interface FormControl extends AbstractControl {
    * It is `any` because the value is untyped.
```
I don't think this comment is necessary anymore 😃 
```suggestion
```

dylhunn replied:
Thanks!

mary-poppins commented:
You can preview f161700 at https://pr43834-f161700.ngbuilds.io/.

dylhunn commented:
merge-assistance: merge separately, Tuesday morning
